File: src/main/java/me/penguin/moneytoken/Objects/CashNote.java
package me.penguin.moneytoken.Objects;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import me.Penguin.SuperPenguinCore.API.PenguinEconomyAPI;
import me.Penguin.SuperPenguinCore.Items.Held.Redeemable;
import me.Penguin.SuperPenguinCore.Util.MIB;
import me.Penguin.SuperPenguinCore.Util.NumberFormatter;
import me.Penguin.SuperPenguinCore.Util.m;
import me.Penguin.SuperPenguinCore.Util.u;

public class CashNote extends Redeemable {
	
	private long amount;
	private String ownerName;
	
	public CashNote(long amount, String ownerName) {
		this.amount = amount;
		this.ownerName = ownerName;
	}
	
	@Override
	public void OnRedeem(ItemStack item, Player player) {
		long amount = getAmount(item);
		PenguinEconomyAPI.deposit(player, amount);
		removeOne(item);
		m.sendPrefixed(player, "&aRedeemed a cash note worth $" + u.dc(amount));
	}

	@Override
	public ItemStack getItem() {
		String format = NumberFormatter.getFormatted(amount);
		if (format.contains(".")) {
			String[] parts = format.split("\\.");
			if (parts[1].length() > 1) parts[1] = parts[1].substring(0, 1) + parts[1].substring(parts[1].length()-1);
			format = String.join(".", parts);
		}
		return new MIB(Material.PAPER)
				.setName("&aCash Note")
				.addLores("&6Worth: $" + u.dc(amount) + (amount >= 1000 ? " ($" + format + ")" : ""))
				.addLores("&6Withdrawn By: &e" + ownerName)
				.addLores(" ")
				.addLores("&7Right-Click to Redeem")
				.setLocname("cashnote:" + amount)
				.build();				
	}

	@Override
	public boolean isItem(ItemStack item) {
		return hasLocname(item) && getLocname(item).contains("cashnote:");
	}
	
	public long getAmount(ItemStack item) {
		return Long.parseLong(getLocname(item).split(":")[1]);
	}
	
	

}

File: src/main/java/com/github/mlgpenguin/moneytokens/items/CashNote.kt
package com.github.mlgpenguin.moneytokens.items

import com.github.supergluelib.customitem.CustomItem
import com.github.supergluelib.foundation.extensions.format
import com.github.supergluelib.foundation.extensions.removeOne
import com.github.supergluelib.foundation.extensions.send
import com.github.supergluelib.foundation.util.ItemBuilder
import com.github.supergluelib.hooks.Hooks
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta

class CashNote(val amount: Int, val withdrawnBy: String): CustomItem() {

    override fun fromItemStack(item: ItemStack, meta: ItemMeta, id: String?) = id!!.split(":").let { CashNote(it[1].toInt(), it[2]) }
    override fun isItem(item: ItemStack, meta: ItemMeta, id: String?) = id!!.startsWith("cashnote:")
    override fun getItem() = ItemBuilder(Material.PAPER, "&aCash Note")
        .addLore("&7Worth: &a$$amount", "&7Withdrawn By: &a$withdrawnBy", "", "&7Right-Click to redeem")
        .identifier("cashnote:$amount:$withdrawnBy")
        .build()

    override fun onRightClickAir(player: Player, item: ItemStack, event: PlayerInteractEvent) {
        Hooks.economy?.deposit(player, amount)
            ?: throw NullPointerException("Economy is null! ${player.name} attempted to redeem a cash note but failed")

        item.removeOne()
        player.send("&7Redeemed &a$${amount.format()}")
    }

}